frame/node-authorization/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+        /// will be updated and keep the existing additional connection unchanged.
+        ///
+        /// May only be called from `T::AddOrigin`.
+        ///
+        /// - `node`: identifier of the node.
+        #[weight = (T::WeightInfo::add_well_known_node(), DispatchClass::Operational)]
+        pub fn add_well_known_node(origin, node: NodePublicKey, owner: T::AccountId) {
+            T::AddOrigin::ensure_origin(origin)?;
+
+            let mut nodes = WellKnownNodes::get();
+            ensure!(nodes.len() < T::MaxWellKnownNodes::get() as usize, Error::<T>::TooManyNodes);
+
+            let location = nodes.binary_search(&node).err().ok_or(Error::<T>::AlreadyJoined)?;
+            nodes.insert(location, node.clone());
+            
+            WellKnownNodes::put(&nodes);
+            <Owners<T>>::insert(&node, &owner);
+
+            Self::deposit_event(RawEvent::NodeAdded(node, owner));
+        }
+
+        /// Remove a node from the set of well known nodes. The ownership and additional
+        /// connections of the node will also be removed.
+        ///
+        /// May only be called from `T::RemoveOrigin`.
+        ///
+        /// - `node`: identifier of the node.
+        #[weight = (T::WeightInfo::remove_well_known_node(), DispatchClass::Operational)]
+        pub fn remove_well_known_node(origin, node: NodePublicKey) {
+            T::RemoveOrigin::ensure_origin(origin)?;
+
+            let mut nodes = WellKnownNodes::get();
+
+            let location = nodes.binary_search(&node).ok().ok_or(Error::<T>::NotExist)?;
+            nodes.remove(location);
+            
+            WellKnownNodes::put(&nodes);
+            <Owners<T>>::remove(&node);
+            AdditionalConnections::remove(&node);
+
+            Self::deposit_event(RawEvent::NodeRemoved(node));
+        }
+
+        /// Swap a well known node to another. Both the ownership and additional connections
+        /// stay untouched.
+        ///
+        /// May only be called from `T::SwapOrigin`.
+        ///
+        /// - `remove`: the node which will be moved out from the list.
+        /// - `add`: the node which will be put in the list.
+        #[weight = (T::WeightInfo::swap_well_known_node(), DispatchClass::Operational)]
+        pub fn swap_well_known_node(origin, remove: NodePublicKey, add: NodePublicKey) {
+            T::SwapOrigin::ensure_origin(origin)?;
+
+            if remove == add { return Ok(()) }
+            
+            let mut nodes = WellKnownNodes::get();
+            let remove_location = nodes.binary_search(&remove).ok().ok_or(Error::<T>::NotExist)?;
+            nodes.remove(remove_location);
+            let add_location = nodes.binary_search(&add).err().ok_or(Error::<T>::AlreadyJoined)?;
+            nodes.insert(add_location, add.clone());
+            
+            WellKnownNodes::put(&nodes);
+            Owners::<T>::swap(&remove, &add);
```

Comment:
This means we keep the owner of `remove` node, no? I think we should rather remove it.

Reply:
The scenario in my mind is that the owner request a swap from an old node to a new one, then approved by a collective.